rpc/providers/proxy.ts

```diff
@@ -3,44 +3,38 @@ import { deadline, deferred } from "../../deps/std/async.ts";
 import { ErrorCtor } from "../../util/mod.ts";
 import * as B from "../Base.ts";
 import { ClientHooks, ParseRawIngressMessageError } from "../common.ts";
+import { WebSocketProxy } from "./web-socket-proxy.ts";
 
 export type ProxyClientHooks<Config_ extends Config<string>> = ClientHooks<Config_, Event>;
 
-export function proxyClient<Config_ extends Config<string>>(
+export async function proxyClient<Config_ extends Config<string>>(
   config: Config_,
   hooks?: ProxyClientHooks<Config_>,
 ): Promise<ProxyClient<Config_> | FailedToOpenConnectionError> {
-  const ws = new WebSocket(config.discoveryValue);
-  const client = new ProxyClient(ws, hooks);
-  ws.addEventListener("error", client.onError);
-  ws.addEventListener("message", client.onMessage);
-  const pending = deferred<ProxyClient<Config_> | FailedToOpenConnectionError>();
-  if (ws.readyState === WebSocket.CONNECTING) {
-    const onOpenError = (e: any) => {
-      console.log({ log: e });
-      clearListeners();
-      pending.resolve(new FailedToOpenConnectionError());
-    };
-    const onOpen = () => {
-      clearListeners();
-      pending.resolve(client);
-    };
-    const clearListeners = () => {
-      ws.removeEventListener("error", onOpenError);
-      ws.removeEventListener("open", onOpen);
-    };
-    ws.addEventListener("error", onOpenError);
-    ws.addEventListener("open", onOpen);
-  } else {
-    pending.resolve(client);
+  const wsIsOpen = deferred();
+  const ws = new WebSocketProxy({
+    webSocketFactory: () => new WebSocket(config.discoveryValue),
+  });
+
+  ws.onopen = () => wsIsOpen.resolve();
+  ws.onerror = () => wsIsOpen.reject();
```

Comment:
This works only for the first connection attempt.

Reply:
This seems to be needed to send messages only when there is an open connection.
If that's the only use case, then, it might be better to handle message sending with a non-open connection in the RPC `Client` base class.
The reason is that the `Client` has enough context to decide what to do when the provider can not send the message.
The client could filter messages and decide for each case something like
- buffer a non-sent messages
- error for messages that can not be re-send
- discard messages

Reply:
The RPC `Client` calls the `Provider.send` assuming that there will be no errors when a message is sent.
I believe that we need to change that assumption and change this signature
```
  /**
   * The provider-specific send implementation
   *
   * @param egressMessage the message you wish to send to the RPC server
   */
  send: (egressMessage: msg.InitMessage<Config_>) => void;
```
to
```
  send: (egressMessage: msg.InitMessage<Config_>) => void | SendError;
```

Then, the `Client` could do
```
  send = (egressMessage: msg.InitMessage<Config_>): void | SendError => {
    this.hooks?.send?.(egressMessage);
    try {
      this.provider.send(egressMessage);
    } catch (e) {
      return new SendError(e);
    }
  };
```

Reply:
Agreed. Is it worthwhile to parameterize the type of the send error independently?

Reply:
I got the suggestion and added `SendError` to `Client` type parameters